[Manage Categories]Synchronize product categories

Ecarrion opened the pull request:
Refs https://github.com/woocommerce/woocommerce-ios/issues/2000

# Why

Following #2043 where an action to retrieve product categories from a remote source was introduced.
This PR modifies that action to store such product categories in our storage layer.

All this efforts are made to power the ProductCategory Selector screen in a following PR.
![Screen Shot 2020-04-02 at 4 47 50 PM](https://user-images.githubusercontent.com/562080/78303135-b75d8d00-7501-11ea-8d17-6afb877f445f.png)


# How
- Adds `SiteID` and `ParentID` to Networking.ProductCategory and `Storage.ProductCategory` - includes xcdatamodel migration to v27

- Update `loadProductCategory` method to not rely on `productID` as categories would be shared through several products.

- Update `ProductCategoryStore` to upset categories in our storage layer upon a successful `loadAllProductCategories` network response.

## Notes
- The big diff is due to the new `xcdatamodel` file.

# Testing Steps
- Open the app with a logged site in a state previous to this PR
- Open the app with a logged site with the code on this PR
- Make sure migration works and no information is lost 😇

Ecarrion commented:
Hi @shiki, please help me with the usual reviewer/label/milestone assignment 🙏

shiki commented on Storage/Storage/Model/WooCommerce.xcdatamodeld/Model 27.xcdatamodel/contents:
```diff
@@ -0,0 +1,554 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
```
I think we also need to update the deletion rules here. 

My understanding of Core Data (especially deletion rules) is very limited so I'm gonna state how I understand the current configuration. 

## My Understanding of the Current Config (Model 27)

### Product to Category

![image](https://user-images.githubusercontent.com/198826/78407591-6bbee800-75c2-11ea-8784-ec71cd05a0f7.png)

The way I understand this is that if the product is deleted, the `categories` should be deleted too. 

### Category to Product

![image](https://user-images.githubusercontent.com/198826/78407670-9b6df000-75c2-11ea-998d-d91c261abba6.png)

I understand this as "If a category is deleted, remove that category from the linked Products' categories". 

## What I Think it Should be

With this new product category synchronization, I think deleting the categories when a product is deleted is not ideal. It should probably be the other way around. However, it's probably also wrong to delete the products if a category is deleted. 

Maybe the rules should be:

- If a product is deleted, the category should be left alone. We don't want to delete the category anymore because we will use it in the picker. 
- If a category is deleted and there are no linked products, it **can** be deleted.
- If a category is deleted and **there are** linked products, the deletion should be **denied**. My concern is that if we use: 
  - `Nullify`: We might end up with a product that is _inaccurate_ (missing category). 
  - `Cascade`: We might delete a product that is currently being edited.

---

Did I understand this correctly? 😅 What do you think about these rules?

@jaclync and @pmusolino Would you mind sharing your thoughts too? 🙂 

@astralbodies Maybe you have some background knowledge related to this?

Ecarrion replied:
I totally forgot about deletion rules 🤦‍♂
After a quick read on them I agree with your assumptions, I've updated them to:
```
Product -> Nullify -> categories
category -> Deny -> products
```
https://github.com/woocommerce/woocommerce-ios/pull/2079/commits/dafea866dc3554c1c49fc76e2d3dfafa1592598b

But I'm also curios to hear about other members thoughts 😇

jaclync replied:
sorry for the delay! from my understanding, I think the deletion rules are like:

- `Product --(to-many)--> Set<ProductCategory>`: Nullify (when a product category is deleted, it just removes itself from the product's to-many relationship, since a product could have multiple categories)
- `ProductCategory --(to-many)--> Set<Product>`: Nullify (when a product is deleted, it just removes itself from the product category's to-many relationship, since many products can share the same category)

lemme know if I missed anything though!

Ecarrion replied:
Hey @jaclync, thanks for your response! 🙇 

> Product --(to-many)--> Set<ProductCategory>: Nullify (when a product category is deleted, it just removes itself from the product's to-many relationship, since a product could have multiple categories)

I've just checked how the web-app works and it does follow that rule. If a product category is deleted, the deletion is allowed and it's removed from the products's category list.

@shiki Are you ok for us to follow that rule?

shiki replied:
@Ecarrion Yeah, that sounds fine to me. The Deny could also be tricky to code around so I'm okay with Nullify for now. 

Let me know if you've made the changes and also how we should migrate to this new structure.

pmusolino replied:
Sorry for the delay, but I see the ping only 2 minutes ago.
I think that what @jaclync said is correct 🙇

shiki commented on Storage/Storage/Model/WooCommerce.xcdatamodeld/Model 27.xcdatamodel/contents:
```diff
@@ -0,0 +1,554 @@
+        <attribute name="parentID" optional="YES" attributeType="Integer 64" defaultValueString="0" usesScalarValueType="YES"/>
+        <attribute name="siteID" optional="YES" attributeType="Integer 64" defaultValueString="0" usesScalarValueType="YES"/>
+        <attribute name="slug" attributeType="String"/>
+        <relationship name="products" optional="YES" toMany="YES" deletionRule="Nullify" destinationEntity="Product" inverseName="categories" inverseEntity="Product"/>
```
This was changed from `maxCount = 1` to `toMany = YES`. That's a change from a single `Product` property to a `Set<Product>`. Were you able to validate that Core Data migrates this successfully? I honestly have no idea. 😅

Ecarrion replied:
According to the documentation I read, CoreData can infer that migration automatically.
And my hand testing seems to confirm that. 

<img width="766" alt="docs" src="https://user-images.githubusercontent.com/562080/78467450-a9129b00-76d2-11ea-9f50-cdebc9ff59d3.png">

shiki commented on Storage/Storage/Model/WooCommerce.xcdatamodeld/Model 27.xcdatamodel/contents:
```diff
@@ -0,0 +1,554 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<model type="com.apple.IDECoreDataModeler.DataModel" documentVersion="1.0" lastSavedToolsVersion="15702" systemVersion="19D76" minimumToolsVersion="Automatic" sourceLanguage="Swift" userDefinedModelVersionIdentifier="">
+    <entity name="Account" representedClassName="Account" syncable="YES">
```
_This is a discussion for the file itself, not line 3_. 

If we change the relationships based on my discussion above, we'd probably also need to do a custom migration. Using Model 26, our database rows for categories look like this:

![image](https://user-images.githubusercontent.com/198826/78410179-d541f500-75c8-11ea-94dd-7362323d7155.png)

Notice that we have a lot of duplicates. We'll probably have to consolidate that if the product → categories relationship is no longer `Cascade Delete`.

Ecarrion replied:
I guess this is because each product had its own category as they weren't shared, right? 

I wonder if it's good enough to delete duplicated-orphan categories when doing `SynchonizeCategories` action? 🤔

shiki replied:
I'd rather not have a "purge" code that is active all the time. I think it's better if we have some sort of migration that only happens when the user has upgraded the app. 

Would you mind creating an issue for the plan for the migration? And then we can decide the priority for it. It can probably be done later after the UI changes as long as we don't forget to remove orphaned records.

shiki commented on Yosemite/Yosemite/Stores/ProductCategoryStore.swift:
```diff
@@ -32,19 +36,64 @@ public final class ProductCategoryStore: Store {
         let remote = ProductCategoriesRemote(network: network)
 
-        remote.loadAllProductCategories(for: siteID, pageNumber: pageNumber, pageSize: pageSize) { (productCategories, error) in
+        remote.loadAllProductCategories(for: siteID, pageNumber: pageNumber, pageSize: pageSize) { [weak self] (productCategories, error) in
```
Works 👍 

![image](https://user-images.githubusercontent.com/198826/78411146-ee987080-75cb-11ea-89ba-9a61752708c7.png)

Ecarrion commented on Yosemite/Yosemite/Stores/ProductCategoryStore.swift:
```diff
@@ -32,19 +36,64 @@ public final class ProductCategoryStore: Store {
+    }
+}
+
+private extension ProductCategoryStore {
```
I thought(to myself) that it should be in the same extension, however after looking at some similar clases like `ProductReviewStore` and `ShippingClassStore` I decided to add it in a separate extension to be online with the style of those.

I think it was done to isolate methods that reach the storage layer directly.

Let me know if you prefer them on the same extension.

Ecarrion commented on Yosemite/Yosemite/Stores/ProductCategoryStore.swift:
```diff
@@ -32,19 +36,64 @@ public final class ProductCategoryStore: Store {
+                                       siteID: Int64) {
+        // Upserts the ProductCategory models from the read-only version
+        for readOnlyProductCategory in readOnlyProductCategories {
+            let storageProductCategory: Storage.ProductCategory = {
+                if let storedCategory = storage.loadProductCategory(siteID: siteID, categoryID: readOnlyProductCategory.categoryID) {
+                    return storedCategory
+                }
+                return storage.insertNewObject(ofType: Storage.ProductCategory.self)
+            }()
+            storageProductCategory.update(with: readOnlyProductCategory)
```
> I feel like a method for the individual ProductCategory would be more useful.

Hmmm 🤔, I guess it may be more useful or would made more sense if the class semantic was all around handling a individual category, however everything seems to be around a group of categories.  However I have no problem on go either way!

----

> Anyway. How did you feel when writing these? Any suggestions? 
> I'd really like these DB update methods outside of the *Store classes but I'm not sure yet what architectural pattern to introduce for that

Do you mean, to not do a `insert` directly in these classes? 
It does feel a bit risky that anyone can insert/delete/update to the underlying store and mess with the data integrity.

What I do normally is to have a `DataStore` type that sits in the outer layer of the `Storage.Framework` and exposes a single way that consumers(`Yosemite` in this case) can interact with a given data type. 

Let's suppose that a product can never be without a category. A `DataStore` would handle that integrity
```swift
final class ProductCatgoriesDataStore {
    func remove(category: ProductCategory) {
        let products = category.products
        storage.remove(category)

        for product in products {
            if product.categories.isEmpty {
                 let defaultCategory = storage.findCategory(id: "uncategorized")
                 product.categories.add(defaultCategory)
                 storage.update(product)
            }
        }
    }
}
```

That way, we stop providing a direct link to our underlying store.

Ecarrion commented on Networking/Networking/Model/Product/ProductCategory.swift:
```diff
@@ -5,29 +5,41 @@ import Foundation
         let container = try decoder.container(keyedBy: CodingKeys.self)
 
         let categoryID = try container.decode(Int64.self, forKey: .categoryID)
+        // Some product endpoints don't include the parent category ID
```
Oh yeah, good point!
I've added some code to prevent it and a test for it! https://github.com/woocommerce/woocommerce-ios/pull/2079/commits/6edc40065647a50b02dcc12e11c31f07d538ea93

shiki replied:
Great! Thanks for adding the test too. That's awesome!

Ecarrion commented on Networking/NetworkingTests/Mapper/ProductCategoyListMapperTests.swift:
```diff
@@ -11,6 +14,7 @@ final class ProductCategoryListMapperTests: XCTestCase {
 
         let firstProductCategory = productCategories[0]
         XCTAssertEqual(firstProductCategory.categoryID, 104)
+        XCTAssertEqual(firstProductCategory.siteID, dummySiteID)
```
Done https://github.com/woocommerce/woocommerce-ios/pull/2079/commits/d0f26a578ad9718092f858345427fa68f67e155a

Ecarrion commented on Yosemite/Yosemite/Stores/ProductStore.swift:
```diff
@@ -379,13 +379,10 @@ private extension ProductStore {
-            if let existingStorageCategory = storage.loadProductCategory(siteID: siteID,
-                                                                         productID: productID,
-                                                                         categoryID: readOnlyCategory.categoryID) {
+            if let existingStorageCategory = storage.loadProductCategory(siteID: siteID, categoryID: readOnlyCategory.categoryID) {
```
Yeah, I guess there is no need to remove the leftovers now as we would need them for the category selector screen. Good spot!

https://github.com/woocommerce/woocommerce-ios/pull/2079/commits/ade9688b52023660a971f04eb7cad23e6890e7f6

Ecarrion commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Product+ProductFormTests.swift:
```diff
@@ -5,6 +5,8 @@ import Yosemite
 
 class Product_ProductFormTests: XCTestCase {
 
+    let sampleSiteID: Int64 = 109
```
Fixed https://github.com/woocommerce/woocommerce-ios/pull/2079/commits/3793177606357b6cbd368091724936d0bfc22477

Ecarrion commented on Yosemite/YosemiteTests/Stores/ProductCategoryStoreTests.swift:
```diff
@@ -47,58 +66,64 @@ final class ProductCategoryStoreTests: XCTestCase {
         // Given a stubed product-categories network response
```
Nice spot! 
Added test https://github.com/woocommerce/woocommerce-ios/pull/2079/commits/a1d1fbadd050629d910486ddcebea6a90d1fc583

shiki commented on Networking/Networking/Model/Product/ProductCategory.swift:
```diff
@@ -58,3 +73,21 @@ extension ProductCategory: Comparable {
+// MARK: - Update functions
+//
+extension ProductCategory {
+    public func updateWith(parentID: Int64) -> ProductCategory {
```
This would probably be deprecated when we have that `copy()` pattern. 😄 For now, would you mind following the method name convention used in [`ProductUpdater`](https://github.com/woocommerce/woocommerce-ios/blob/develop/Yosemite/Yosemite/Model/Updaters/ProductUpdater.swift)?

```suggestion
    public func parentIDUpdated(parentID: Int64) -> ProductCategory {
```

The `updateWith()` doesn't sound like a new instance will be created. 😐

Ecarrion replied:
Fixed https://github.com/woocommerce/woocommerce-ios/pull/2079/commits/dadb2f45367f4bf64979fd6c0f972564aa400d23

Ecarrion commented:
Hi @shiki, as per our conversation I've:
- Left the `deny` option on the `category -> Product` delete rule. https://github.com/woocommerce/woocommerce-ios/pull/2079/commits/dafea866dc3554c1c49fc76e2d3dfafa1592598b
- Added code to delete unused(no product association) categories on `synchronizeProductCategoriesAction`  https://github.com/woocommerce/woocommerce-ios/pull/2079/commits/cf7348455a46d6d58faf83e20e1bb9c417672df1
- Added a migration to delete product categories without a siteID(all stored categories so far) using a mapping model. https://github.com/woocommerce/woocommerce-ios/pull/2079/commits/a61817ecdf12da286f7bc9e962c20a09fc5cf104

shiki commented on Storage/Storage/Model/WooCommerce.xcdatamodeld/Model 27.xcdatamodel/contents:
```diff
@@ -0,0 +1,554 @@
+        <attribute name="categoryID" attributeType="Integer 64" defaultValueString="0" usesScalarValueType="YES"/>
+        <attribute name="name" attributeType="String"/>
+        <attribute name="parentID" optional="YES" attributeType="Integer 64" defaultValueString="0" usesScalarValueType="YES"/>
+        <attribute name="siteID" optional="YES" attributeType="Integer 64" defaultValueString="0" usesScalarValueType="YES"/>
```
⚠️  Since the migration deletes all the existing `ProductCategory` rows, we now have the opportunity to make this `siteID` **required**. What do you think?

Ecarrion replied:
Oh yeah! Nice catch!

Ecarrion commented on Storage/Storage/Tools/StorageType+Deletions.swift:
```diff
@@ -40,4 +40,18 @@ public extension StorageType {
+
+    /// Deletes all of the stored Product Categories that don't have an active product relationship
+    ///
+    func deleteUnusedProductCategories(siteID: Int64) {
```
Done!

Ecarrion commented on Storage/Storage/Model/MIGRATIONS.md:
```diff
@@ -2,6 +2,12 @@
+- @ecarrion 2020-03-30
+- Update `ProductCategory`'s `product` relationship to `products`
+- Add `siteID` and `parentID` to  `ProductCategory` entity 
+- Used mapping model: `WooCommerceModelV26toV27.xcmappingmodel` to remove product categories without `siteID`
```
Exactly!

Ecarrion commented:
Changes applied @shiki
